### PR TITLE
Add custom headers while code generation

### PR DESCRIPTION
**Pull request by paule96**
### Issues

*This pull request fixes issue #xxx.*
#1315 

### Description

Add support to add custom headers for the code generation. Now you can set the property `CustomHttpHeaders` in the `Configuration` class with your custom headers. The code generation use this for the HTTP requests to your OData API.

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

**Comment by paule96** [comment]
There are some open points:

- The user that use the generated code must add an EventHandler to his client that set the headers for his requests. (maybe this is okay)
- There are no tests for TT template so I don't write some tests
- I implement my own basic json array pharser. maybe we can do this better...

**Comment by paule96** [comment]
And I can't add the label "Docs Needed"

**Comment by paule96** [comment]
@raheph and @mikepizzo

**Comment by mikepizzo** [comment]
Are the two .json files in .vscode part of the fix, or did they get picked up accidentally?

**Comment by paule96** [comment]
@mikepizzo it isn't part of the fix but it would make development with vscode much easier. I can remove this here. Should I?

If yes, I create a separate PR for this changes 😊

**Comment by mikepizzo** [comment]
Yes; please remove it from this PR and open a separate one.  I'm all for working well with vscode.

Thanks!

**Comment by paule96** [comment]
done

**Comment by paule96** [comment]
@mikepizzo can approve this?

File: src/CodeGen/ODataT4CodeGenerator.tt [hunk 1]
```diff
@@ -27,7 +27,9 @@ public static class Configuration
 
 	// This flag indicates whether to ignore unexpected elements and attributes in the metadata document and generate
 	// the client code if any. The value must be set to true or false.
-	public const bool IgnoreUnexpectedElementsAndAttributes = true;
+	public const bool IgnoreUnexpectedElementsAndAttributes = true;	
+	// Here you can set custom http headers like this: [{'headername': 'headervalue'}, {'headername': 'headervalue' }]
```
Comment (KanishManuja-MS) [inline src/CodeGen/ODataT4CodeGenerator.tt]: It would be nice if you can add to the comment for the configuration about when this will be consumed by T4.

**Comment by mikepizzo** [comment]
I spent some time thinking about the last point.  I was trying to come up with a simpler syntax for parsing a collection of name/value pairs, but the fact that virtually any character can be in the header value makes that difficult.  So I think a json-like syntax is probably fine, although I would shy away from describing it as a "json" syntax; it may look similar to json, but we don't want to imply that it's *actually* json.

For example, I don't know if we should have the key/values be (single) quoted, as I'm not sure it buys us much, and single quotes are valid within a header-value.

Note that, in HTTP, (double) quotes within header values are meaningful.  Most values are not quoted, for example:

     Accept:application/json;odata.metadata=full,application; q=0.4

But quotes surrounding an etag *are* sent as part of the header value:

     If-Match: "xyzzy"
or
     If-Match: W/"xyzzy"

While, for a star request they are not:
     If-Match: *

So we should allow escaped double=quotes (\") within the key-value.

Also, the current simple validation logic wouldn't handle things like curly braces and colons within the header value.  Braces may be okay, but there are header values that use colons, i.e., for date/time or URLs, that would get messed up by the validation logic. i.e.;

    If-Modified-Since: Sat, 29 Oct 1994 19:43:31 GMT

or

   Location: [http://www.w3.org/pub/WWW/People.html](http://www.w3.org/pub/WWW/People.html)

So, maybe we simplify it to:
"{key:value},{key:value}"

and allow escaping double-quotes, say that we don't support curly braces within a header value, and remove/refine the validation?

---
In reply to: [443873749](https://github.com/OData/odata.net/pull/1348#issuecomment-443873749) [](ancestors = 443873749)

File: src/CodeGen/ODataT4CodeGenerator.tt [hunk 2]
```diff
@@ -27,7 +27,9 @@ public static class Configuration
 
 	// This flag indicates whether to ignore unexpected elements and attributes in the metadata document and generate
 	// the client code if any. The value must be set to true or false.
-	public const bool IgnoreUnexpectedElementsAndAttributes = true;
+	public const bool IgnoreUnexpectedElementsAndAttributes = true;	
+	// Here you can set custom http headers like this: [{'headername': 'headervalue'}, {'headername': 'headervalue' }]
+	public const string CustomHttpHeaders = "[]";
```
Comment (mikepizzo) [inline src/CodeGen/ODataT4CodeGenerator.tt]: Might be easier to simplify this to remove the square brackets and single quotes. Note that we need to support escaped double-quotes within the string for key-values containing double quotes.

File: src/CodeGen/ODataT4CodeGenerator.ttinclude [hunk 3]
```diff
@@ -317,6 +326,45 @@ public void ValidateAndSetIgnoreUnexpectedElementsAndAttributesFromString(string
...
+        !stringValue.Contains("{") || !stringValue.Contains("}") || 
+        // Check if the count of the keys are the same as the count of values
+        (stringValue.Count(d => d == ':') != stringValue.Count(d => d == '{')) ||
+        // Check all objects of a '{' and a '}'
+        (stringValue.Count(d => d == '{') != stringValue.Count(d => d == '}')) ||
+        // Check all values and keys are strings
+        (stringValue.Count(d => d == ':')*4 != stringValue.Count(d => d == '\''))
+        ){
+        throw new ArgumentException("Please check your custom http headers string format!");
+    }
```
Comment (mikepizzo) [inline src/CodeGen/ODataT4CodeGenerator.ttinclude]: I don't know that these validations are valid; for example, a colon is a valid (and often used) character within a key-value.

**Comment by paule96** [comment]
Sorry I was offline because holidays etc. I'm now start working again on this. :) Happy new year all.

**Comment by paule96** [comment]
with my last commit, I hope I have implemented the logic @mikepizzo describe in his [comment](https://github.com/OData/odata.net/pull/1348#issuecomment-447116560).

Please review this.

File: src/CodeGen/ODataT4CodeGenerator.tt (hunk 1 above)
Comment (paule96) [reply src/CodeGen/ODataT4CodeGenerator.tt]: I think this is done.

**Comment by paule96** [comment]
@KanishManuja-MS can you show me some example tests and help me a little bit. I'm a little bit lost in the test cases and I don't know how to write tests for T4 templates. 🆘

**Comment by KanishManuja-MS** [comment]
You can take a look [here ](https://github.com/OData/odata.net/blob/eb0fc9a1e5a7ef58de205f99818e446b6e5c85cd/test/FunctionalTests/Tests/DataServices/UnitTests/DesignT4UnitTests/ODataT4CodeGeneratorTestDescriptors.cs#L183)and write your cases. ODataT4CodeGeneratorTestDescriptors.cs

**Comment by paule96** [comment]
I have found [this](http://earlz.net/view/2012/11/21/0346/how-to-unit-test-t4-code-generators). But this is a complete rewrite of the ttinclude file only for test....

**Comment by paule96** [comment]
Can anybody post the build output here? Because I can't build the solution from master local. :(

**Comment by xuzhg** [comment]
@paule96 

```
C:\BuildAgent\_work\36\s\tools\PoliCheck\RunPoliCheck.ps1 : PoliCheck failed for target C:\BuildAgent\_work\36\s\src. 
For details, please check this result file on build machine: 
C:\BuildAgent\_work\36\a\PoliCheck\master\poli_result_src.xml: section <Result TotalOccurences=...>.
At line:1 char:1
+ . 'C:\BuildAgent\_work\36\s\tools\PoliCheck\RunPoliCheck.ps1' -folder ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,RunPoliCheck.ps1
 


Process completed with exit code 1 and had 1 error(s) written to the error stream.
```

and
```
2019-01-15T14:06:17.5181269Z Failed   TestTheValidationForASingleHeader
2019-01-15T14:06:17.5181725Z ##[error]Error Message:
2019-01-15T14:06:17.5182222Z ##[error]   Assert.AreEqual failed. Expected:<1>. Actual:<0>. 
2019-01-15T14:06:17.5182702Z ##[error]Stack Trace:
2019-01-15T14:06:17.5183430Z ##[error]   at Microsoft.OData.Client.Design.T4.UnitTests.CodeGenerationContextTest.TestTheValidationForASingleHeader() in C:\BuildAgent\_work\36\s\test\FunctionalTests\Tests\DataServices\UnitTests\DesignT4UnitTests\CodeGenerationContextTest.cs:line 305
2019-01-15T14:06:17.5184254Z 
2019-01-15T14:06:18.8814344Z ##[error]Error Message:
2019-01-15T14:06:18.8815495Z ##[error]   Assert.AreEqual failed. Expected:<2>. Actual:<0>. 
2019-01-15T14:06:18.8960660Z ##[error]Stack Trace:
2019-01-15T14:06:18.8969743Z ##[error]   at Microsoft.OData.Client.Design.T4.UnitTests.CodeGenerationContextTest.TestTheValidationForMultipleHeaders() in C:\BuildAgent\_work\36\s\test\FunctionalTests\Tests\DataServices\UnitTests\DesignT4UnitTests\CodeGenerationContextTest.cs:line 322
2019-01-15T14:06:18.8970924Z ##[error]Error Message:
2019-01-15T14:06:18.8971535Z ##[error]   Assert.Fail failed. 
2019-01-15T14:06:18.8972652Z ##[error]Stack Trace:
2019-01-15T14:06:18.8973562Z ##[error]   at Microsoft.OData.Client.Design.T4.UnitTests.CodeGenerationContextTest.TestTheValidationErrorForMissingColonInHeaders() in C:\BuildAgent\_work\36\s\test\FunctionalTests\Tests\DataServices\UnitTests\DesignT4UnitTests\CodeGenerationContextTest.cs:line 349
2019-01-15T14:06:18.8974436Z ##[error]Error Message:
2019-01-15T14:06:18.8974963Z ##[error]   Assert.Fail failed. 
2019-01-15T14:06:18.8975834Z ##[error]Stack Trace:
2019-01-15T14:06:18.8977186Z ##[error]   at Microsoft.OData.Client.Design.T4.UnitTests.CodeGenerationContextTest.TestTheValidationErrorForMissingCurlyBreaketsAtTheEndInHeaders() in C:\BuildAgent\_work\36\s\test\FunctionalTests\Tests\DataServices\UnitTests\DesignT4UnitTests\CodeGenerationContextTest.cs:line 369
2019-01-15T14:06:18.8978067Z ##[error]Error Message:
2019-01-15T14:06:18.8978583Z ##[error]   Assert.Fail failed. 
2019-01-15T14:06:18.8979086Z ##[error]Stack Trace:
2019-01-15T14:06:18.8979946Z ##[error]   at Microsoft.OData.Client.Design.T4.UnitTests.CodeGenerationContextTest.TestTheValidationErrorForMissingCurlyBreaketsAtTheBeginInHeaders() in C:\BuildAgent\_work\36\s\test\FunctionalTests\Tests\DataServices\UnitTests\DesignT4UnitTests\CodeGenerationContextTest.cs:line 389
2019-01-15T14:06:18.8981279Z ##[error]Error Message:
2019-01-15T14:06:18.8981870Z ##[error]   Assert.Fail failed. 
2019-01-15T14:06:18.8982382Z ##[error]Stack Trace:
2019-01-15T14:06:18.8983209Z ##[error]   at Microsoft.OData.Client.Design.T4.UnitTests.CodeGenerationContextTest.TestTheValidationErrorForMissingCommaBetweenHeaders() in C:\BuildAgent\_work\36\s\test\FunctionalTests\Tests\DataServices\UnitTests\DesignT4UnitTests\CodeGenerationContextTest.cs:line 412
2019-01-15T14:06:18.8984425Z Failed   TestTheValidationForMultipleHeaders
2019-01-15T14:06:18.8984801Z 
2019-01-15T14:06:18.9053873Z Failed   TestTheValidationErrorForMissingColonInHeaders
2019-01-15T14:06:18.9073344Z 
2019-01-15T14:06:18.9079139Z Failed   TestTheValidationErrorForMissingCurlyBreaketsAtTheEndInHeaders
2019-01-15T14:06:18.9087157Z 
2019-01-15T14:06:18.9087917Z Failed   TestTheValidationErrorForMissingCurlyBreaketsAtTheBeginInHeaders
2019-01-15T14:06:18.9088460Z 
2019-01-15T14:06:18.9089071Z Failed   TestTheValidationErrorForMissingCommaBetweenHeaders
`

**Comment by KanishManuja-MS** [comment]
>           Can anybody post the build output here? Because I can't build the solution from master local. :(

image

Not all these errors may be because of the changes, you can try and run them locally. Along with that the Policheck is failing as well.

**Comment by KenitoInc** [comment]
This PR was abandoned. Another PR dealing with this issue is under review

**Pull request closed without merging**